Allow generators to impl Clone/Copy

canndrew opened the pull request:
This PR allows generators to implement `Clone`/`Copy` if all their upvars and all their locals which are held across a yield implement `Clone`/`Copy`. This only applies to non-`static` generators, and so excludes generators which contain self-references and the generators that `async` blocks/fns desugar to,

r? @oli-obk

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5617853839?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/0eacb42a9a4ac28dbac5ff98b416f992862b4bf1/checks/5617853839/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0787
Found 504 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/compiler/rustc_feature/src/active.rs:400: feature generator_clone is not sorted by feature name (should be between fn_align and generators)
tidy error: /checkout/compiler/rustc_feature/src/active.rs:400: no tracking issue for feature generator_clone
Expected a gate test for the feature 'generator_clone'.
tidy error: Found 1 features without a gate test.
Hint: create a failing test file named 'feature-gate-generator_clone.rs'
      in the 'ui' test suite, with its failures due to
      missing usage of `#![feature(generator_clone)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-generator_clone line to the test file.
some tidy checks failed
```

</details>

oli-obk commented on compiler/rustc_mir_transform/src/shim.rs:
```diff
@@ -459,49 +462,111 @@ impl<'tcx> CloneShimBuilder<'tcx> {
 
             let dest_field = self.tcx.mk_place_field(dest, field, ity);
 
-            // #(2i + 1) is the cleanup block for the previous clone operation
```
it took me a bit to untangle this function by looking at its callers and tracing its logic. Please (re-)add some more comments, especially around the arguments.

oli-obk commented on compiler/rustc_mir_transform/src/shim.rs:
```diff
@@ -459,49 +462,111 @@ impl<'tcx> CloneShimBuilder<'tcx> {
+            let variant_index = VariantIdx::new(index);
+            let dest = self.tcx.mk_place_downcast_unnamed(dest, variant_index);
+            let src = self.tcx.mk_place_downcast_unnamed(src, variant_index);
+            let start_block = self.block_index_offset(0);
+            let clone_block = self.block_index_offset(1);
+            cases.push((index as u128, start_block));
```
if you move the `cases.push` and `let start_block` after the `self.block` below, you don't need to use `self.block_index_offset(0)` here as you can get the id from the `self.block`

canndrew replied:
I've added some comments. Let me know what you think.

oli-obk commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -1918,6 +1917,39 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                let resolved_witness = self.infcx.shallow_resolve(substs.as_generator().witness());
+                if {
+                    matches!(resolved_upvars.kind(), ty::Infer(ty::TyVar(_))) ||
+                    matches!(resolved_witness.kind(), ty::Infer(ty::TyVar(_)))
```
there is a `is_ty_var()` method that is exactly this match internally.

oli-obk commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -1918,6 +1917,39 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                    Ambiguous
+                } else {
+                    let mut all = substs.as_generator().upvar_tys().collect::<Vec<_>>();
+                    all.push(substs.as_generator().witness());
```
instead of collecting an iterator and then pushing, chain the iterator with the extra element and collect that.

oli-obk commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -1918,6 +1917,39 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+            ty::GeneratorWitness(binder) => {
+                let tys = binder.no_bound_vars().unwrap();
+                let mut iter = tys.iter();
+                loop {
```
please use a for loop and a return instead of a break, I found this hard to read

oli-obk reviewed (changes requested):
I'd like to see some tests on various kinds of async blocks, too. Even if we're just testing the diagnostics that way.

canndrew commented:
Sorry for the delay. I've done the requested changes.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5959543465?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/811b2b9ee98b197013836564ba03701380a55472/checks/5959543465/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
...................................................iii.................................. 12848/12918
......................................................................
failures:

---- [ui] src/test/ui/generator/clone-impl.rs stdout ----

133    |
133    |
134 LL | fn check_copy<T: Copy>(_x: &T) {}
135    |                  ^^^^ required by this bound in `check_copy`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+ help: consider annotating `NonClone` with `#[derive(Copy)]`
+    |
+ LL | #[derive(Copy)]
136 
136 
137 error[E0277]: the trait bound `NonClone: Clone` is not satisfied in `[generator@$DIR/clone-impl.rs:62:25: 65:6]`
138   --> $DIR/clone-impl.rs:68:5
157    |
157    |
158 LL | fn check_clone<T: Clone>(_x: &T) {}
159    |                   ^^^^^ required by this bound in `check_clone`
+ help: consider annotating `NonClone` with `#[derive(Clone)]`
+ LL | #[derive(Clone)]
+    |
160 
161 error: aborting due to 6 previous errors
---
To only update this specific test, also pass `--test-args generator/clone-impl.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/generator/clone-impl.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/clone-impl" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generator/clone-impl/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0277]: the trait bound `Vec<u32>: Copy` is not satisfied in `[generator@/checkout/src/test/ui/generator/clone-impl.rs:36:23: 41:6]`
  --> /checkout/src/test/ui/generator/clone-impl.rs:42:5
   |
LL |       let gen_clone_0 = move || {
   |  _______________________-
LL | |         let v = vec!['a'];
LL | |         yield;
LL | |         drop(v);
LL | |         drop(clonable_0);
LL | |     };
   | |_____- within this `[generator@/checkout/src/test/ui/generator/clone-impl.rs:36:23: 41:6]`
LL |       check_copy(&gen_clone_0);
   |       ^^^^^^^^^^ within `[generator@/checkout/src/test/ui/generator/clone-impl.rs:36:23: 41:6]`, the trait `Copy` is not implemented for `Vec<u32>`
note: captured value does not implement `Copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:40:14
   |
   |
LL |         drop(clonable_0);
   |              ^^^^^^^^^^ has type `Vec<u32>` which does not implement `Copy`
note: required by a bound in `check_copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:72:18
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `Vec<char>: Copy` is not satisfied in `[generator@/checkout/src/test/ui/generator/clone-impl.rs:36:23: 41:6]`
  --> /checkout/src/test/ui/generator/clone-impl.rs:42:5
   |
LL |       let gen_clone_0 = move || {
   |  _______________________-
LL | |         let v = vec!['a'];
LL | |         yield;
LL | |         drop(v);
LL | |         drop(clonable_0);
LL | |     };
   | |_____- within this `[generator@/checkout/src/test/ui/generator/clone-impl.rs:36:23: 41:6]`
LL |       check_copy(&gen_clone_0);
   |       ^^^^^^^^^^ within `[generator@/checkout/src/test/ui/generator/clone-impl.rs:36:23: 41:6]`, the trait `Copy` is not implemented for `Vec<char>`
   |
note: generator does not implement `Copy` as this value is used across a yield
  --> /checkout/src/test/ui/generator/clone-impl.rs:38:9
   |
LL |         let v = vec!['a'];
   |             - has type `Vec<char>` which does not implement `Copy`
LL |         yield;
   |         ^^^^^ yield occurs here, with `v` maybe used later
LL |     };
LL |     };
   |     - `v` is later dropped here
note: required by a bound in `check_copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:72:18
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `Vec<u32>: Copy` is not satisfied in `[generator@/checkout/src/test/ui/generator/clone-impl.rs:46:23: 57:6]`
  --> /checkout/src/test/ui/generator/clone-impl.rs:58:5
   |
LL |       let gen_clone_1 = move || {
   |  _______________________-
LL | |         let v = vec!['a'];
LL | |         /*
LL | |         let n = NonClone;
...  |
LL | |         drop(clonable_1);
LL | |     };
   | |_____- within this `[generator@/checkout/src/test/ui/generator/clone-impl.rs:46:23: 57:6]`
LL |       check_copy(&gen_clone_1);
   |       ^^^^^^^^^^ within `[generator@/checkout/src/test/ui/generator/clone-impl.rs:46:23: 57:6]`, the trait `Copy` is not implemented for `Vec<u32>`
note: captured value does not implement `Copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:56:14
   |
   |
LL |         drop(clonable_1);
   |              ^^^^^^^^^^ has type `Vec<u32>` which does not implement `Copy`
note: required by a bound in `check_copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:72:18
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `Vec<char>: Copy` is not satisfied in `[generator@/checkout/src/test/ui/generator/clone-impl.rs:46:23: 57:6]`
  --> /checkout/src/test/ui/generator/clone-impl.rs:58:5
   |
LL |       let gen_clone_1 = move || {
   |  _______________________-
LL | |         let v = vec!['a'];
LL | |         /*
LL | |         let n = NonClone;
...  |
LL | |         drop(clonable_1);
LL | |     };
   | |_____- within this `[generator@/checkout/src/test/ui/generator/clone-impl.rs:46:23: 57:6]`
LL |       check_copy(&gen_clone_1);
   |       ^^^^^^^^^^ within `[generator@/checkout/src/test/ui/generator/clone-impl.rs:46:23: 57:6]`, the trait `Copy` is not implemented for `Vec<char>`
   |
note: generator does not implement `Copy` as this value is used across a yield
  --> /checkout/src/test/ui/generator/clone-impl.rs:52:9
   |
LL |         let v = vec!['a'];
   |             - has type `Vec<char>` which does not implement `Copy`
LL |         yield;
LL |         yield;
   |         ^^^^^ yield occurs here, with `v` maybe used later
LL |     };
LL |     };
   |     - `v` is later dropped here
note: required by a bound in `check_copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:72:18
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`

error[E0277]: the trait bound `NonClone: Copy` is not satisfied in `[generator@/checkout/src/test/ui/generator/clone-impl.rs:62:25: 65:6]`
  --> /checkout/src/test/ui/generator/clone-impl.rs:66:5
LL |       let gen_non_clone = move || {
   |  _________________________-
LL | |         yield;
LL | |         yield;
LL | |         drop(non_clonable);
LL | |     };
   | |_____- within this `[generator@/checkout/src/test/ui/generator/clone-impl.rs:62:25: 65:6]`
LL |       check_copy(&gen_non_clone);
   |       ^^^^^^^^^^ within `[generator@/checkout/src/test/ui/generator/clone-impl.rs:62:25: 65:6]`, the trait `Copy` is not implemented for `NonClone`
note: captured value does not implement `Copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:64:14
   |
   |
LL |         drop(non_clonable);
   |              ^^^^^^^^^^^^ has type `NonClone` which does not implement `Copy`
note: required by a bound in `check_copy`
  --> /checkout/src/test/ui/generator/clone-impl.rs:72:18
   |
LL | fn check_copy<T: Copy>(_x: &T) {}
   |                  ^^^^ required by this bound in `check_copy`
help: consider annotating `NonClone` with `#[derive(Copy)]`
   |
LL | #[derive(Copy)]


error[E0277]: the trait bound `NonClone: Clone` is not satisfied in `[generator@/checkout/src/test/ui/generator/clone-impl.rs:62:25: 65:6]`
  --> /checkout/src/test/ui/generator/clone-impl.rs:68:5
LL |       let gen_non_clone = move || {
   |  _________________________-
LL | |         yield;
LL | |         yield;
LL | |         drop(non_clonable);
LL | |     };
   | |_____- within this `[generator@/checkout/src/test/ui/generator/clone-impl.rs:62:25: 65:6]`
...
LL |       check_clone(&gen_non_clone);
   |       ^^^^^^^^^^^ within `[generator@/checkout/src/test/ui/generator/clone-impl.rs:62:25: 65:6]`, the trait `Clone` is not implemented for `NonClone`
note: captured value does not implement `Clone`
  --> /checkout/src/test/ui/generator/clone-impl.rs:64:14
   |
   |
LL |         drop(non_clonable);
   |              ^^^^^^^^^^^^ has type `NonClone` which does not implement `Clone`
note: required by a bound in `check_clone`
  --> /checkout/src/test/ui/generator/clone-impl.rs:73:19
   |
LL | fn check_clone<T: Clone>(_x: &T) {}
   |                   ^^^^^ required by this bound in `check_clone`
help: consider annotating `NonClone` with `#[derive(Clone)]`
LL | #[derive(Clone)]
   |

error: aborting due to 6 previous errors
```

</details>

canndrew commented:
I've also added a test which checks that feature(generator_clone) doesn't interfere with the clonability of `async` blocks and doesn't effect the error message of trying to clone an `async` block. And I've also rebased off of master in order to fix the ui tests which were failing due to minor changes in diagnostics.

oli-obk commented on compiler/rustc_trait_selection/src/traits/select/mod.rs:
```diff
@@ -1897,6 +1896,43 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                        // Not yet resolved.
+                        Ambiguous
+                    } else {
+                        let all = substs
+                            .as_generator()
+                            .upvar_tys()
+                            .chain(iter::once(substs.as_generator().witness()))
+                            .collect::<Vec<_>>();
```
Not due to this PR, but I really wish we had ways to track where an inference var got resolved and then bubble this span up here. In the current design there is no way to point at the local that causes the generator to stop being clone/copy, all we can do is keep pointing at the entire generator :(

bors commented:
:umbrella: The latest upstream changes (presumably #96883) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
Oof I forgot about this. Changed the label so it shows up in my queue

oli-obk commented:
ok, did another pass. This lgtm now. r=me after a rebase

oli-obk commented:
@rustbot author

JohnCSimon commented:
> ok, did another pass. This lgtm now. r=me after a rebase


ping from triage:
@canndrew 
Can you please rebase so we can move forward?

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

JohnCSimon commented:
@canndrew 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
